lib/msf/core/exe/segment_injector.rb

```diff
@@ -58,27 +58,26 @@ def create_thread_stub
 
     def create_thread_stub_x64
       <<-EOS
+        sub rsp, 38h
+
         mov rcx, hook_libname
-        sub rsp, 30h
         mov rax, iat_LoadLibraryA
         call [rax]
-        add rsp, 30h
 
         mov rdx, hook_funcname
         mov rcx, rax
-        sub rsp, 30h
         mov rax, iat_GetProcAddress
         call [rax]
-        add rsp, 30h
 
-        push 0
-        push 0
-        mov r9, 0
+        xor ecx, ecx
+        mov qword ptr [rsp+28h], rcx
+        mov qword ptr [rsp+20h], rcx
```

Comment:
I'll admit that this change confuses me a little. I would expect to see two `0` values at `[rsp]` and `[rsp+8]` to represent parameters 5 and 6. But that's not what this is doing.

Can you please explain why you're writing zeros to the shadow space rather than at the top of the stack?

Reply:
> why you're writing zeros to the shadow space

That's exactly what the previous code did, and this fixes it. Here's one of the Google Images results for "shadow stack x64", feel free to read more about it: https://modexp.files.wordpress.com/2017/01/x64_hs.png?w=640

Reply:
I can confirm some x86_64 calling conventions I have used put params 5+ on the stack where they would have been under an all-stack (x86-style) convention. The [MS docs on calling convention](https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019) support this as well:
> The x64 Application Binary Interface (ABI) uses a four-register fast-call calling convention by default. Space is allocated on the call stack as a shadow store for callees to save those registers. ... Any parameters beyond the first four must be stored on the stack after the shadow store before the call.